File: forms/issue-tracker/front-end-form-with-svelte/src/lib/engine/composite-grader.test.ts
import { describe, expect, it } from 'vitest';
import { gradeIssue } from './composite-grader';
import type { IssueTrackerAssessment, RawScores } from './types';

function blankAssessment(scores: Partial<RawScores> = {}): IssueTrackerAssessment {
	return {
		reporter: {
			reporterName: '',
			reporterEmail: '',
			reporterRole: '',
			reportedAt: '',
			discoveredAt: '',
			issueCategory: '',
			environment: '',
			systemName: '',
			component: '',
			customerOrProjectTag: '',
			externalReference: '',
		},
		cc: { ccSummary: '', ccLongDescription: '', ccReportedByName: '', ccReportedVia: '' },
		pt: {
			ptDiscovererName: '',
			ptAffectedUsersCount: null,
			ptAffectedUserGroups: '',
			ptAssignees: '',
			ptStakeholdersToInform: '',
			ptObservers: '',
		},
		sx: {
			sxExternalSignals: '',
			sxAlertIds: '',
			sxErrorMessages: '',
			sxScreenshotsUrl: '',
			sxLogsUrl: '',
			sxFirstObservedAt: '',
		},
		fx: {
			fxBrokenComponents: '',
			fxFailedServices: '',
			fxStuckProcesses: '',
			fxHardwareFaults: '',
			fxDataCorruption: '',
		},
		hx: {
			hxRelatedIssues: '',
			hxPriorOccurrences: null,
			hxRecentChangeUrl: '',
			hxReferences: '',
			hxTimeline: '',
		},
		ix: {
			ixHypotheses: '',
			ixReproSteps: '',
			ixDiagnosticQueries: '',
			ixTestsRun: '',
			ixBlockingUnknowns: '',
		},
		dx: { dxRootCause: '', dxContributingCauses: '', dxScope: '', dxConfirmed: '' },
		txpx: {
			txMitigationSteps: '',
			txFixPlan: '',
			txWorkaround: '',
			txRollbackPlan: '',
			txCommunicationPlan: '',
			pxExpectedResolutionAt: '',
			pxResidualRisk: '',
			pxMonitoringPlan: '',
			pxRecurrenceLikelihood: '',
			pxLessonsLearned: '',
		},
		scores: {
			scoreByPriorityRank: null,
			scoreBySeverityOfImpact: null,
			scoreByMagnitudeOfDamage: null,
			scoreByHarmGrade: null,
			scoreByFailureCondition: '',
			scoreByMoscowRequirement: null,
			scoreByFrequencyPercent: null,
			...scores,
		},
	};
}

describe('gradeIssue — empty assessment', () => {
	it('grades an empty assessment as low priority with only the composite rule fired', () => {
		const r = gradeIssue(blankAssessment());
		expect(r.compositePriority).toBe('low');
		expect(r.firedRules).toHaveLength(1);
		expect(r.firedRules[0].instrument).toBe('composite');
		expect(r.additionalFlags).toEqual([]);
	});
});

describe('gradeIssue — single-dimension drives composite (max-grade)', () => {
	it('priority rank 1 → critical', () => {
		const r = gradeIssue(blankAssessment({ scoreByPriorityRank: 1 }));
		expect(r.compositePriority).toBe('critical');
	});
	it('severity 5 → critical', () => {
		const r = gradeIssue(blankAssessment({ scoreBySeverityOfImpact: 5 }));
		expect(r.compositePriority).toBe('critical');
	});
	it('magnitude 10 → critical', () => {
		const r = gradeIssue(blankAssessment({ scoreByMagnitudeOfDamage: 10 }));
		expect(r.compositePriority).toBe('critical');
	});
	it('harm 4 → critical', () => {
		const r = gradeIssue(blankAssessment({ scoreByHarmGrade: 4 }));
		expect(r.compositePriority).toBe('critical');
	});
	it('failure A → critical', () => {
		const r = gradeIssue(blankAssessment({ scoreByFailureCondition: 'A' }));
		expect(r.compositePriority).toBe('critical');
	});
	it('moscow 1 → high (not critical)', () => {
		const r = gradeIssue(blankAssessment({ scoreByMoscowRequirement: 1 }));
		expect(r.compositePriority).toBe('high');
	});
	it('frequency 100% → critical', () => {
		const r = gradeIssue(blankAssessment({ scoreByFrequencyPercent: 100 }));
		expect(r.compositePriority).toBe('critical');
	});
	it('failure D + severity 1 + frequency 1% stays low', () => {
		const r = gradeIssue(
			blankAssessment({
				scoreByFailureCondition: 'D',
				scoreBySeverityOfImpact: 1,
				scoreByFrequencyPercent: 1,
			}),
		);
		expect(r.compositePriority).toBe('low');
	});
});

describe('gradeIssue — safety flags', () => {
	it('harm 4 fires the harm-fatal flag', () => {
		const r = gradeIssue(blankAssessment({ scoreByHarmGrade: 4 }));
		expect(r.additionalFlags.map((f) => f.category)).toContain('harm-fatal');
	});
	it('failure A fires the failure-catastrophic flag', () => {
		const r = gradeIssue(blankAssessment({ scoreByFailureCondition: 'A' }));
		expect(r.additionalFlags.map((f) => f.category)).toContain('failure-catastrophic');
	});
	it('severity 5 fires the severity-catastrophic flag', () => {
		const r = gradeIssue(blankAssessment({ scoreBySeverityOfImpact: 5 }));
		expect(r.additionalFlags.map((f) => f.category)).toContain('severity-catastrophic');
	});
	it('magnitude 10 fires the magnitude-total-destruction flag', () => {
		const r = gradeIssue(blankAssessment({ scoreByMagnitudeOfDamage: 10 }));
		expect(r.additionalFlags.map((f) => f.category)).toContain('magnitude-total-destruction');
	});
	it('frequency >= 95 fires the frequency-universal flag', () => {
		const r = gradeIssue(blankAssessment({ scoreByFrequencyPercent: 99 }));
		expect(r.additionalFlags.map((f) => f.category)).toContain('frequency-universal');
	});
	it('clinical-safety + harm 2 fires the regulatory flag', () => {
		const data = blankAssessment({ scoreByHarmGrade: 2 });
		data.reporter.issueCategory = 'clinical-safety';
		const r = gradeIssue(data);
		expect(r.additionalFlags.map((f) => f.category)).toContain('regulatory');
	});
	it('moscow 1 fires the requirement-mandatory flag', () => {
		const r = gradeIssue(blankAssessment({ scoreByMoscowRequirement: 1 }));
		expect(r.additionalFlags.map((f) => f.category)).toContain('requirement-mandatory');
	});
});

describe('gradeIssue — fired rules', () => {
	it('records one rule per non-null instrument plus the composite rule', () => {
		const r = gradeIssue(
			blankAssessment({
				scoreByPriorityRank: 2,
				scoreBySeverityOfImpact: 3,
				scoreByMagnitudeOfDamage: 5,
				scoreByHarmGrade: 1,
				scoreByFailureCondition: 'C',
				scoreByMoscowRequirement: 2,
				scoreByFrequencyPercent: 25,
			}),
		);
		// 7 dimensions + 1 composite
		expect(r.firedRules).toHaveLength(8);
		expect(r.firedRules.map((x) => x.instrument)).toEqual([
			'priority',
			'severity',
			'magnitude',
			'harm',
			'failure',
			'moscow',
			'frequency',
			'composite',
		]);
	});
});

File: forms/issue-tracker/full-stack-with-loco-tera-htmx-alpine/tests/engine/composite_grader_test.rs
use issue_tracker_tera_crate::engine::composite_grader::grade;
use issue_tracker_tera_crate::engine::types::*;

fn empty_data() -> AssessmentData {
    AssessmentData::default()
}

#[test]
fn all_null_scores_collapse_to_low_composite() {
    let result = grade(&empty_data());
    assert_eq!(result.composite_priority, "low");
    // Only the composite "R-COMPOSITE-LOW" rule should fire when every
    // input is null.
    assert!(
        result
            .fired_rules
            .iter()
            .any(|r| r.rule_id == "R-COMPOSITE-LOW")
    );
}

#[test]
fn priority_rank_1_yields_critical_composite() {
    let mut data = empty_data();
    data.scores.score_by_priority_rank = Some(1);
    let result = grade(&data);
    assert_eq!(result.composite_priority, "critical");
    assert!(
        result
            .fired_rules
            .iter()
            .any(|r| r.rule_id == "R-PRIORITY-1" && r.band == "critical")
    );
    assert!(
        result
            .fired_rules
            .iter()
            .any(|r| r.rule_id == "R-COMPOSITE-CRITICAL")
    );
}

#[test]
fn severity_5_drives_critical_composite() {
    let mut data = empty_data();
    data.scores.score_by_severity_of_impact = Some(5);
    let result = grade(&data);
    assert_eq!(result.composite_priority, "critical");
    assert!(
        result
            .fired_rules
            .iter()
            .any(|r| r.rule_id == "R-SEVERITY-5")
    );
}

#[test]
fn magnitude_10_drives_critical_composite() {
    let mut data = empty_data();
    data.scores.score_by_magnitude_of_damage = Some(10);
    let result = grade(&data);
    assert_eq!(result.composite_priority, "critical");
    assert!(
        result
            .fired_rules
            .iter()
            .any(|r| r.rule_id == "R-MAGNITUDE-10")
    );
}

#[test]
fn harm_grade_4_drives_critical_composite() {
    let mut data = empty_data();
    data.scores.score_by_harm_grade = Some(4);
    let result = grade(&data);
    assert_eq!(result.composite_priority, "critical");
    assert!(
        result
            .fired_rules
            .iter()
            .any(|r| r.rule_id == "R-HARM-4")
    );
}

#[test]
fn failure_condition_a_drives_critical_composite() {
    let mut data = empty_data();
    data.scores.score_by_failure_condition = "A".to_string();
    let result = grade(&data);
    assert_eq!(result.composite_priority, "critical");
    assert!(
        result
            .fired_rules
            .iter()
            .any(|r| r.rule_id == "R-FAILURE-A")
    );
}

#[test]
fn moscow_1_drives_high_composite() {
    let mut data = empty_data();
    data.scores.score_by_moscow_requirement = Some(1);
    let result = grade(&data);
    assert_eq!(result.composite_priority, "high");
    assert!(
        result
            .fired_rules
            .iter()
            .any(|r| r.rule_id == "R-MOSCOW-1" && r.band == "high")
    );
}

#[test]
fn frequency_100_drives_critical_composite() {
    let mut data = empty_data();
    data.scores.score_by_frequency_percent = Some(100.0);
    let result = grade(&data);
    assert_eq!(result.composite_priority, "critical");
    // R-FREQUENCY-{round(f)} == "R-FREQUENCY-100"
    assert!(
        result
            .fired_rules
            .iter()
            .any(|r| r.rule_id == "R-FREQUENCY-100")
    );
}

#[test]
fn worst_band_wins_under_max_grade() {
    let mut data = empty_data();
    // priority=4 (low), severity=4 (high), magnitude=2 (low),
    // harm=2 (high), failure=D (low), moscow=3 (low), freq=5 (moderate).
    // The composite should be the worst — "high".
    data.scores.score_by_priority_rank = Some(4);
    data.scores.score_by_severity_of_impact = Some(4);
    data.scores.score_by_magnitude_of_damage = Some(2);
    data.scores.score_by_harm_grade = Some(2);
    data.scores.score_by_failure_condition = "D".to_string();
    data.scores.score_by_moscow_requirement = Some(3);
    data.scores.score_by_frequency_percent = Some(5.0);
    let result = grade(&data);
    assert_eq!(result.composite_priority, "high");
}

#[test]
fn integer_frequency_emits_no_decimal_in_grade_string() {
    let mut data = empty_data();
    data.scores.score_by_frequency_percent = Some(42.0);
    let result = grade(&data);
    let rule = result
        .fired_rules
        .iter()
        .find(|r| r.instrument == "frequency")
        .expect("frequency rule");
    assert_eq!(rule.grade, "42%");
}

#[test]
fn fractional_frequency_keeps_two_decimal_places() {
    let mut data = empty_data();
    data.scores.score_by_frequency_percent = Some(12.5);
    let result = grade(&data);
    let rule = result
        .fired_rules
        .iter()
        .find(|r| r.instrument == "frequency")
        .expect("frequency rule");
    assert_eq!(rule.grade, "12.50%");
}
